### PR TITLE
Loki: strip out invalid options for logs/metrics queries

### LINKED ISSUE #78096: Loki: query were not triaged  in the dashboard  after enabled lokiQuerySplitting for target with `"maxLines": 0`
### What happened?

The dashboard is not loaded, and click on `Query` button, nothing happened.
The json for this panels as
`{
          "datasource": {
            "type": "loki",
            "uid": "cff9e619-5bd3-46f8-8863-56ed36c6bd10"
          },
          "editorMode": "code",
          "expr": "sum (quantile_over_time(0.50, {_namespace_=\"xxx\"}|logfmt|t_site_id=0|cmd_comm=\"a\"|dd_d_de=\"b\"|dd_bf_b=~\"Chrome.*|Safari|Firefox\"|unwrap jbyte[$__interval]))",
          "hide": false,
          "maxLines": 0,
          "queryType": "range",
          "refId": "A"
        },`
image

after removed `"maxLines": 0,` 
The dashboard could be loaded. 
image

### What did you expect to happen?

The dashboard could be loaded, and query can be executed when `"maxLines": 0` were set. 

### Did this work before?

Yes, disable lokiQuerySplitting, the query can be loaded. 

### How do we reproduce it?

1.  In Json Model, append `"maxLines": 0` for the queries. Save it. 
2. Check the dashboard, you can reproduce it. 

### Is the bug inside a dashboard panel?

_No response_

### Environment (with versions)?

Grafana: 10.1.5
OS: MacOS 13.5.1
Browser: Chrome

### Grafana platform?

None

### Datasource(s)?

Loki

### PR DESCRIPTION
**Pull request by gtk-grafana**
**What is this feature?**
Fix for bug introduced when log query options are being evaluated against metric queries.

**Why do we need this feature?**
Users cannot change maxLines after they have written a metric query, since this field is not visible to users, they have the expectation that it will not impact the results of their query. The UX should reflect this assumption.

**Who is this feature for?**
Loki users

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/78096

**Special notes for your reviewer:**
We might want to make this change as well on the backend?

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by matyax — APPROVED** [review]
👏 👏

File: public/app/plugins/datasource/loki/datasource.ts [hunk 1]
```diff
@@ -294,7 +295,8 @@ export class LokiDatasource
   query(request: DataQueryRequest<LokiQuery>): Observable<DataQueryResponse> {
     const queries = request.targets
       .map(getNormalizedLokiQuery) // used to "fix" the deprecated `.queryType` prop
-      .map((q) => ({ ...q, maxLines: q.maxLines ?? this.maxLines }));
+      .map((q) => ({ ...q, maxLines: q.maxLines ?? this.maxLines }))
+      .map(removeInvalidOptions);
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/datasource.ts:299]: Why isn't this in `getNormalizedLokiQuery` and move `.map((q) => ({ ...q, maxLines: q.maxLines ?? this.maxLines })` as `getNormalizedLokiQuery`  seems like a perfect place for this.

**Review by ivanahuckova — COMMENTED** [review]
I feel like this is an issue with `lokiQuerySplitting` and this is not a correct fix, because even though we are not going to be sending `maxLines` / `step` to backend if the queries are running though browser, we still can't rely on it in backend as alerting queries are not going to go trough this code. So even though we are fixing the issue, it creates difference between alerting/non-alerting queries that we can't rely on.

**Comment by matyax** [comment]
> we still can't rely on it in backend as alerting queries are not going to go trough this code

That's a great point. Let's focus on the query splitting part of this. I can help @gtk-grafana

**Comment by gtk-grafana** [comment]
So I am not seeing this issue in the alerting UI, or on the alerts executed on the backend, on this branch or on main
image
image

I'm going to take a look at the backend today and see if we can find out what is breaking there, but keep in mind only the frontend knows if a **request** is a metrics or logs query, the backend doesn't have any lezer parsing.
https://raintank-corp.slack.com/archives/C04LFDM7PMY/p1702328207403089

**Comment by gtk-grafana** [comment]
So it looks like the query is not sent to the backend when the maxLines is set to 0, must be something downstream from this change that is not emitting the request if max lines is 0.

**Comment by ivanahuckova** [comment]
@gtk-grafana I thought that the issue is related to what @matyax mentioned:

> I see. maxLines is used for query splitting when querying logs, and it requires a value > 0 to properly work, so there's no workaround for that now.

So I would expect that we check for `maxLines`  only for log queries in query splitting. And query splitting is on frontend.

File: public/app/plugins/datasource/loki/datasource.ts (hunk 1 above)
Comment (gtk-grafana) [reply public/app/plugins/datasource/loki/datasource.ts]: I went ahead and made this change

**Comment by gtk-grafana** [comment]
Gotcha, yeah I found a much simpler solution in the query splitting, closing this one out!

**Comment by gtk-grafana** [comment]
Moved to https://github.com/grafana/grafana/pull/79409

**Pull request closed without merging**